drivers/timer/nrf_grtc_timer.c

```diff
@@ -0,0 +1,571 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/irq.h>
+#include <zephyr/drivers/timer/system_timer.h>
+#include <zephyr/drivers/timer/nrf_grtc_timer.h>
+#include <nrfx_grtc.h>
+#include <zephyr/sys/math_extras.h>
+
+#if CONFIG_SYSTEM_CLOCK_WAIT
+#include <zephyr/drivers/clock_control/nrf_clock_control.h>
+#endif
+
+#define CHAN_COUNT     NRFX_GRTC_CONFIG_NUM_OF_CC_CHANNELS
+#define EXT_CHAN_COUNT (CHAN_COUNT - 1)
+/* The reset value of waketime is 1, which doesn't seem to work.
+ * It's being looked into, but for the time being use 4.
+ * Timeout must always be higher than waketime, so setting that to 5.
+ */
+#define WAKETIME       (4)
+#define TIMEOUT        (WAKETIME + 1)
+
+#define GRTC_NODE DT_NODELABEL(grtc)
+
+#ifndef GRTC_SYSCOUNTERL_VALUE_Msk
+#define GRTC_SYSCOUNTERL_VALUE_Msk GRTC_SYSCOUNTER_SYSCOUNTERL_VALUE_Msk
+#endif
+
+#ifndef GRTC_SYSCOUNTERH_VALUE_Msk
+#define GRTC_SYSCOUNTERH_VALUE_Msk GRTC_SYSCOUNTER_SYSCOUNTERH_VALUE_Msk
+#endif
+
+#define MAX_CC_LATCH_WAIT_TIME_US 77
+
+#define CYC_PER_TICK                                                                               \
+	((uint64_t)sys_clock_hw_cycles_per_sec() / (uint64_t)CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+
+#define COUNTER_SPAN (GRTC_SYSCOUNTERL_VALUE_Msk | ((uint64_t)GRTC_SYSCOUNTERH_VALUE_Msk << 32))
+#define MAX_TICKS                                                                                  \
+	(((COUNTER_SPAN / CYC_PER_TICK) > INT_MAX) ? INT_MAX : (COUNTER_SPAN / CYC_PER_TICK))
+
+#define MAX_CYCLES (MAX_TICKS * CYC_PER_TICK)
+
+/* The maximum SYSCOUNTERVALID settling time equals 1x32k cycles + 20x1MHz cycles. */
+#define GRTC_SYSCOUNTERVALID_SETTLE_MAX_TIME_US 51
+
+#if defined(CONFIG_TEST)
+const int32_t z_sys_timer_irq_for_test = DT_IRQN(GRTC_NODE);
+#endif
+
+static void sys_clock_timeout_handler(int32_t id, uint64_t cc_val, void *p_context);
+
+static struct k_spinlock lock;
+static uint64_t last_count;
+static atomic_t int_mask;
+static uint8_t ext_channels_allocated;
+static nrfx_grtc_channel_t system_clock_channel_data = {
+	.handler = sys_clock_timeout_handler,
+	.p_context = NULL,
+	.channel = (uint8_t)-1,
+};
+
+#define IS_CHANNEL_ALLOWED_ASSERT(chan)                                                            \
+	__ASSERT_NO_MSG((NRFX_GRTC_CONFIG_ALLOWED_CC_CHANNELS_MASK & (1UL << (chan))) &&           \
+			((chan) != system_clock_channel_data.channel))
+
+static inline void grtc_active_set(void)
+{
+#if defined(NRF_GRTC_HAS_SYSCOUNTER_ARRAY) && (NRF_GRTC_HAS_SYSCOUNTER_ARRAY == 1)
+	nrfy_grtc_sys_counter_active_set(NRF_GRTC, true);
+	while (!nrfy_grtc_sys_conter_ready_check(NRF_GRTC)) {
+	}
+#else
+	nrfy_grtc_sys_counter_active_state_request_set(NRF_GRTC, true);
+	k_busy_wait(GRTC_SYSCOUNTERVALID_SETTLE_MAX_TIME_US);
```

Comment:
in which scenarios will enter this 50us busywait branch?

Reply:
CC @adamkondraciuk

Reply:
unanswered